src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+        SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction = refSource.getReferenceWindowFunction();
+        for (GATKRead r : shard.reads) {
+            SimpleInterval readRefs = referenceWindowFunction.apply(r);
+            start = Math.min(readRefs.getStart(), start);
+            end = Math.max(readRefs.getEnd(), end);
+        }
+
+        if (start==Integer.MAX_VALUE) {
+            // there are no reads in this shard, so we're going to remove it
+            return null;
+        }
+
+        SimpleInterval refInterval = new SimpleInterval(shard.interval.getContig(), start, end);
+        ReferenceBases refBases;
+        try {
+            refBases = refSource.getReferenceBases(null, refInterval);
+        } catch (IOException x) {
+            throw new GATKException("Unable to read the reference");
+        }
+
+        ArrayList<ReadContextData> readContext = new ArrayList<>();
+        for (GATKRead r : shard.reads) {
+            SimpleInterval readInterval = new SimpleInterval(r);
+            ArrayList<Variant> variantsOverlappingThisRead = shard.variantsOverlapping(readInterval);
+            // we pass all the bases. That's better because this way it's just a shared
+            // pointer instead of being an array copy. Downstream processing is fine with having
+            // extra bases (it expects a few, actually).
+            readContext.add(new ReadContextData(refBases, variantsOverlappingThisRead));
+        }
+        return shard.withReadContext(readContext);
+    }
+
+    protected static final class SubdivideAndFillReadsIterator implements Iterator<ContextShard>, Serializable {
+        private static final long serialVersionUID = 1L;
+        private static final Logger log = LogManager.getLogger(SubdivideAndFillReadsIterator.class);
+        private final transient SamReader reader;
+        private final transient SAMRecordIterator query;
+        private final ContextShard shard;
+        private final String bam;
+        private final ReadFilter optFilter;
+        private final int lastValidPos;
+        private final int firstValidPos;
+        private final List<SimpleInterval> subshards;
+        private int currentSubShardIndex;
+        private ArrayList<GATKRead> readsSoFar = new ArrayList<>();
+        private Storage.Objects storageClient = null;
+        private SimpleInterval currentSubShard;
+        private ContextShard nextOutput = null;
+        private boolean readerClosed = false;
+        private int outputCount;
+
+        public SubdivideAndFillReadsIterator(String bam, AuthHolder auth, int outputShardSize, int margin, final ReadFilter optFilter, ContextShard shard) throws IOException, GeneralSecurityException, ClassNotFoundException {
+            this.bam = bam;
+            this.shard = shard;
+            this.optFilter = optFilter;
+            lastValidPos = shard.interval.getEnd() + margin;
+            firstValidPos = shard.interval.getStart() - margin;
```

Comment:
We only use it to check if reads come in too early - it's fine if it's negative since no read will start before 1 anyways. But I've changed it so we don't have to go through this whole thought process when reading the code, and instead it just looks right.